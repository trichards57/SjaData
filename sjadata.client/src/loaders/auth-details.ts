export const scopes = [
  "email",
  "profile",
  "a984d5ce-d914-47d0-b690-1bcf084eb829/.default",
];

export const tenantId = "91d037fb-4714-4fe8-b084-68c083b8193f";
export const clientId = "a984d5ce-d914-47d0-b690-1bcf084eb829";
export const redirectUri = "https://localhost:5173";
