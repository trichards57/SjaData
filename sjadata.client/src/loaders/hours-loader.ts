import { useMsal } from "@azure/msal-react";
import {
  QueryClient,
  queryOptions,
  useSuspenseQuery,
} from "@tanstack/react-query";

interface HoursCount {
  counts: Partial<Record<AreaLabel, string>>;
  lastUpdate: string;
}

export type Trust =
  | "NEAS"
  | "NWAS"
  | "WMAS"
  | "EMAS"
  | "EEAST"
  | "LAS"
  | "SECAMB"
  | "SWAST"
  | "SCAS"
  | "YAS"
  | "WAST"
  | "SAS"
  | "NIAS"
  | "IWAS";

export type Region = "NE" | "NW" | "WM" | "EM" | "EOE" | "LON" | "SE" | "SW";

export type AreaLabel = Region | Trust;

export interface ParsedHoursCount {
  counts: Partial<Record<AreaLabel, number>>;
  lastUpdate: Date | undefined;
}

function hoursOptions(token: string, date?: Date, future?: boolean) {
  let dateString: string | undefined = undefined;
  if (date) {
    dateString = date.toISOString().split("T")[0];
  }

  const loader = hoursLoader(token);

  return queryOptions({
    queryKey: ["hours", dateString, future],
    queryFn: () => loader(dateString, future),
  });
}

export function useHoursCount(date?: Date, future: boolean = false) {
  const msal = useMsal();

  return useSuspenseQuery(
    hoursOptions(msal.accounts[0].idToken ?? "", date, future)
  );
}

export function preloadHoursCount(
  queryClient: QueryClient,
  token: string,
  date?: Date,
  future: boolean = false
) {
  queryClient.ensureQueryData(hoursOptions(token, date, future));
}

function hoursLoader(token: string) {
  const authHeader = `Bearer ${token}`;

  return async (date?: string, future: boolean = false) => {
    let uri = date
      ? `/api/hours/count?date=${date}&dateType=m&api-version=1.0`
      : "/api/hours/count?api-version=1.0";

    if (future) {
      uri = `${uri}&future=true`;
    }

    const res = await fetch(uri, {
      headers: {
        Authorization: authHeader,
      },
    });

    if (!res.ok) throw new Error("Failed to load hours details.");

    const data = (await res.json()) as HoursCount;

    const parsedData: ParsedHoursCount = {
      counts: {},
      lastUpdate:
        data.lastUpdate === "0001-01-01T00:00:00+00:00"
          ? undefined
          : new Date(data.lastUpdate),
    };

    for (const key of Object.keys(data.counts) as AreaLabel[]) {
      const value = data.counts[key];

      if (!value) {
        continue;
      }

      let daysString: string = "0";
      let time: string = "";

      if (value.includes(".")) {
        [daysString, time] = value.split(".");
      } else {
        time = value;
      }

      const parts = time.split(":");
      const days = parseInt(daysString, 10);
      const hours = parseInt(parts[0], 10);
      const minutes = parseInt(parts[1], 10);

      parsedData.counts[key] = days * 24 + hours + minutes / 60;
    }

    return parsedData as Readonly<ParsedHoursCount>;
  };
}
