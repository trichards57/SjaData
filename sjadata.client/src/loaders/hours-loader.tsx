interface HoursCount {
  counts: Record<string, string>;
  lastUpdate: string;
}

export interface ParsedHoursCount {
  counts: Record<string, number>;
  lastUpdate: Date;
}

export default async function hoursLoader(date?: Date) {
  const uri = date
    ? `/api/hours/count?date=${date.toISOString().split("T")[0]}&dateType=Month`
    : "/api/hours/count";

  const res = await fetch(uri);

  if (!res.ok) throw new Error("Failed to load hours details.");

  const data = (await res.json()) as HoursCount;

  const parsedData: ParsedHoursCount = {
    counts: {},
    lastUpdate: new Date(data.lastUpdate),
  };

  for (const [key, value] of Object.entries(data.counts)) {
    const [daysString, time] = value.split(".");
    const parts = time.split(":");
    const days = parseInt(daysString, 10);
    const hours = parseInt(parts[0], 10);
    const minutes = parseInt(parts[1], 10);

    parsedData.counts[key] = Math.round(days * 24 + hours + minutes / 60);
  }

  return parsedData as Readonly<ParsedHoursCount>;
}
