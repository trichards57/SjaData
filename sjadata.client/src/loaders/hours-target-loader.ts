import { useMsal } from "@azure/msal-react";
import {
  QueryClient,
  queryOptions,
  useSuspenseQuery,
} from "@tanstack/react-query";
import loader from "./loader";
import { IPublicClientApplication } from "@azure/msal-browser";

interface TargetHours {
  target: number;
}

function hoursTargetOptions(app: IPublicClientApplication, date?: Date) {
  let dateString: string | undefined = undefined;
  if (date) {
    dateString = date.toISOString().split("T")[0];
  }

  const uri = date
    ? `/api/hours/target?date=${dateString}&dateType=m&api-version=1.0`
    : "/api/hours/target?api-version=1.0";

  const load = loader<TargetHours>(app, uri);

  // const loader = hoursTargetLoader(token);

  return queryOptions({
    queryKey: ["hours-target", dateString],
    queryFn: load,
  });
}

export function useHoursTarget(date?: Date) {
  const msal = useMsal();
  return useSuspenseQuery(hoursTargetOptions(msal.instance, date));
}

export function preloadHoursTargetCount(
  queryClient: QueryClient,
  app: IPublicClientApplication,
  date?: Date
) {
  queryClient.ensureQueryData(hoursTargetOptions(app, date));
}
