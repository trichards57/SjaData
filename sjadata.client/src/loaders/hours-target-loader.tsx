import { useMsal } from "@azure/msal-react";
import {
  QueryClient,
  queryOptions,
  useSuspenseQuery,
} from "@tanstack/react-query";

interface TargetHours {
  target: number;
}

function hoursTargetOptions(token: string, date?: Date) {
  let dateString: string | undefined = undefined;
  if (date) {
    dateString = date.toISOString().split("T")[0];
  }

  const loader = hoursTargetLoader(token);

  return queryOptions({
    queryKey: ["hours-target", dateString],
    queryFn: () => loader(dateString),
  });
}

export function useHoursTarget(date?: Date) {
  const msal = useMsal();
  return useSuspenseQuery(
    hoursTargetOptions(msal.accounts[0].idToken ?? "", date)
  );
}

export function preloadHoursTargetCount(
  queryClient: QueryClient,
  token: string,
  date?: Date
) {
  queryClient.ensureQueryData(hoursTargetOptions(token, date));
}

function hoursTargetLoader(token: string) {
  const authHeader = `Bearer ${token}`;
  console.log(authHeader);

  return async (date?: string) => {
    const uri = date
      ? `/api/hours/target?date=${date}&dateType=m&api-version=1.0`
      : "/api/hours/target?api-version=1.0";

    const res = await fetch(uri, {
      headers: {
        Authorization: authHeader,
      },
    });

    if (!res.ok) throw new Error("Failed to load hours details.");

    const data = (await res.json()) as TargetHours;

    return data.target;
  };
}
