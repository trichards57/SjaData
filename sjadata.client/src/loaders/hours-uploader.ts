import { useMsal } from "@azure/msal-react";

export function useHoursUploader() {
  const msal = useMsal();

  return async (file: File) => {
    const tokenRes = await msal.instance.acquireTokenSilent({
      scopes: ["User.Read"],
      account: msal.instance.getAccount({
        tenantId: "91d037fb-4714-4fe8-b084-68c083b8193f",
      })!,
    });

    const authHeader = `Bearer ${tokenRes.idToken}`;

    const formData = new FormData();
    formData.append("file", file);

    const res = await fetch("/api/hours?api-version=1.0", {
      method: "POST",
      headers: {
        Authorization: authHeader,
      },
      body: formData,
    });

    return res.ok ? ((await res.json()) as { count: number }) : undefined;
  };
}
