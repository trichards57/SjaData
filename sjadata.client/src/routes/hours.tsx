import { createFileRoute } from "@tanstack/react-router";
import hoursLoader, { ParsedHoursCount } from "../loaders/hours-loader";
import hoursTargetLoader from "../loaders/hours-target-loader";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck, faMinus, faPlus, faXmark } from "@fortawesome/free-solid-svg-icons";
import { Loading } from "../components/loading";
import { useState } from "react";

interface HoursProps {
  month: Readonly<ParsedHoursCount>;
  ytd: Readonly<ParsedHoursCount>;
  target: number;
}

export const Route = createFileRoute("/hours")({
  component: function Component() {
    return <Hours {...Route.useLoaderData()} />;
  },
  pendingComponent: Loading,
  loader: async () => ({
    month: await hoursLoader(new Date()),
    ytd: await hoursLoader(),
    target: await hoursTargetLoader(new Date()),
  }),
});

export function Hours({ ytd, month, target }: HoursProps) {
  const [expandNHSE, setExpandNHSE] = useState(false);

  const hoursTotal = Math.round(
    Object.values(ytd.counts).reduce((acc, val) => acc + val, 0)
  );

  const monthTotal = Math.round(
    Object.values(month.counts).reduce((acc, val) => acc + val, 0)
  );

  return (
    <>
      <section>
        <h2>Hours</h2>
        <div className="link-boxes">
          <div className="hours-box month">
            <div>This Month</div>
            <div>{monthTotal}</div>
          </div>
          <div className="hours-box ytd">
            <div>Year to Date</div>
            <div>{hoursTotal}</div>
          </div>
          <div className="hours-box target">
            <div>
              NHSE Target
              <a className="see" href="#target-note">
                [*]
              </a>
            </div>
            <div>{target}</div>
          </div>
        </div>
        <p>
          <a id="target-note">*</a> Target is shown as people-hours, not
          crew-hours (and so is double what we bill to NHSE).
        </p>
      </section>
      <section>
        <h3>Filter</h3>
        <ul className="area-list">
          <li className="area-section" onClick={() => setExpandNHSE((s) => !s)}>
            {expandNHSE ? (
              <FontAwesomeIcon icon={faMinus} />
            ) : (
              <FontAwesomeIcon icon={faPlus} />
            )}{" "}
            NHSE Contract
          </li>
          <ul className={"area-section-list" + (expandNHSE ? " expanded" : "")}>
            <li>
              <FontAwesomeIcon icon={faCheck} /> North East Ambulance Service
            </li>
            <li>
              <FontAwesomeIcon icon={faXmark} /> North West Ambulance Service
            </li>
            <li>West Midlands Ambulance Service</li>
            <li>East Midlands Ambulance Service</li>
            <li>East of England Ambulance Service</li>
            <li>South West Ambulance Service</li>
            <li>South Central Ambulance Service</li>
            <li>South East Coast Ambulance Service</li>
            <li>London Ambulance Service</li>
            <li>Yorkshire Ambulance Service</li>
            <li>Isle of Wight Ambulance Service</li>
          </ul>
        </ul>
      </section>
      <p className="last-update">Data last updated : {ytd.lastUpdate?.toLocaleString() ?? "No Data"} </p>
    </>
  );
}
