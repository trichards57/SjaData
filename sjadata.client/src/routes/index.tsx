import { createFileRoute, Link } from '@tanstack/react-router'

export const Route = createFileRoute('/')({
    component: Index,
})

function Index() {
    return (
        <>
            <h2>Home</h2>
            <div className="link-boxes">
                <Link className="link-box" to="/hours">Hours</Link>
                <Link className="link-box" to="/patients">Patients</Link>
            </div>
        </>
    )
}
