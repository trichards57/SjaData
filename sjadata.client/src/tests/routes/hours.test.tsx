import { describe, expect, it } from "vitest";
import { render } from "@testing-library/react";
import { Hours } from "../../routes/hours";
import { ParsedHoursCount } from "../../loaders/hours-loader";

const testMonthData: ParsedHoursCount = {
  counts: {
    NE: 11.3,
    NW: 12.3,
  },
  lastUpdate: new Date("2022-01-02T00:00:00Z"),
};

const testYearData: ParsedHoursCount = {
  counts: {
    NE: 153,
    NW: 143,
  },
  lastUpdate: new Date("2022-01-02T00:00:00Z"),
};

const testTarget = 2000;

describe("Hours Page", () => {
  it("renders data correctly", async () => {
    const { asFragment } = render(
      <Hours month={testMonthData} ytd={testYearData} target={testTarget} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
