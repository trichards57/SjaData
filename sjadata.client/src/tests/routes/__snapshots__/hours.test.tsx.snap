// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hours Page > renders data correctly 1`] = `
<DocumentFragment>
  <section>
    <h2>
      Hours
    </h2>
    <div
      class="link-boxes"
    >
      <div
        class="hours-box month"
      >
        <div>
          This Month
        </div>
        <div>
          24
        </div>
      </div>
      <div
        class="hours-box ytd"
      >
        <div>
          Year to Date
        </div>
        <div>
          296
        </div>
      </div>
      <div
        class="hours-box target"
      >
        <div>
          NHSE Target
          <a
            class="see"
            href="#target-note"
          >
            [*]
          </a>
        </div>
        <div>
          2000
        </div>
      </div>
    </div>
    <p>
      <a
        id="target-note"
      >
        *
      </a>
       Target is shown as people-hours, not crew-hours (and so is double what we bill to NHSE).
    </p>
  </section>
  <section>
    <h3>
      Filter
    </h3>
    <ul
      class="area-list"
    >
      <li>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-plus "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 144L48 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l144 0 0 144c0 17.7 14.3 32 32 32s32-14.3 32-32l0-144 144 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-144 0 0-144z"
            fill="currentColor"
          />
        </svg>
        NHSE Contract
      </li>
      <ul>
        <li>
          North East Ambulance Service
        </li>
        <li>
          North West Ambulance Service
        </li>
        <li>
          West Midlands Ambulance Service
        </li>
        <li>
          East Midlands Ambulance Service
        </li>
        <li>
          East of England Ambulance Service
        </li>
        <li>
          South West Ambulance Service
        </li>
        <li>
          South Central Ambulance Service
        </li>
        <li>
          South East Coast Ambulance Service
        </li>
        <li>
          London Ambulance Service
        </li>
        <li>
          Yorkshire Ambulance Service
        </li>
        <li>
          Isle of Wight Ambulance Service
        </li>
      </ul>
    </ul>
  </section>
  <p>
    Data last updated : 02/01/2022, 00:00:00 
  </p>
</DocumentFragment>
`;
